src/orthophotomap/forest_segmentation.py

```diff
@@ -65,30 +48,53 @@ def mask(self, rgb_image: np.ndarray):
+        '''
+        Change shadows only image to mask. Shadows between trees left unchanged
+        :param rgb_shadowed: Shadows only image, normalized
+        :return: mask with the shadows cut out
+        '''
+        r_gray = cv2.cvtColor(rgb_shadowed, cv2.COLOR_RGB2GRAY)
+        r_gray = apply_brightness_contrast(r_gray, self.shadow_brightness_correction, self.shadow_contrast_correction)
+        entr = entropy(r_gray, disk(self.shadow_entropy_size))
         entr = (entr - entr.min()) / (entr.max() - entr.min())
         entr = cv2.GaussianBlur(entr, (5, 5), cv2.BORDER_DEFAULT)
-        kernel = np.ones((15, 15), np.uint8)
-        ret, mask_r = cv2.threshold(entr, 0.65, 1, cv2.THRESH_BINARY)
+        kernel = np.ones((self.shadow_opening_size, self.shadow_opening_size), np.uint8)
+        ret, mask_r = cv2.threshold(entr, self.shadow_threshold, 1, cv2.THRESH_BINARY)
         mask_r = cv2.morphologyEx(mask_r, cv2.MORPH_OPEN, kernel)
-        mask_r = np.uint8(mask_r)
-
-        # Masking the image
-        img_new = cv2.bitwise_and(rgb_image, rgb_image, mask=mask_r)
+        return np.uint8(mask_r)
 
-        # Final Entropy filtering, removing excess material
-        l_channel = cv2.cvtColor(img_new, cv2.COLOR_RGB2GRAY)
-        l_channel = self._apply_brightness_contrast(l_channel, 64, 30)
-        entr_new = entropy(l_channel, disk(45))
+    def filter_entropy_by_color(self, rgb_img):
+        '''
+        Entropy filter the image, to leave textures similiar to trees
+        :param rgb_img: rgb image valued 0-255, 3 channels
+        :return: mask of everything without tree texture
+        '''
+        l_channel = cv2.cvtColor(rgb_img, cv2.COLOR_RGB2GRAY)
+        l_channel = apply_brightness_contrast(l_channel, self.rgb_brightness_correction, self.rgb_contrast_correction)
+        entr_new = entropy(l_channel, disk(self.rgb_entropy_size))
         entr_new = (entr_new - entr_new.min()) / (entr_new.max() - entr_new.min())
-        kernel = np.ones((15, 15), np.uint8)
-        ret, mask_r_new = cv2.threshold(entr_new, 0.65, 1, cv2.THRESH_BINARY)
+        kernel = np.ones((self.rgb_opening_size, self.rgb_opening_size), np.uint8)
+        ret, mask_r_new = cv2.threshold(entr_new, self.rgb_threshold, 1, cv2.THRESH_BINARY)
         mask_r_new = cv2.morphologyEx(mask_r_new, cv2.MORPH_OPEN, kernel)
-        mask_r_new = np.uint8(mask_r_new)
+        return np.uint8(mask_r_new)
+
+    def mask(self, rgb_image: np.ndarray):
+        '''
+        Mask the provided image, removing most of the non-forest areas
+        :param rgb_image: Channels last rgb image of values 0-255
+        :return: single channel mask of 0/1 values of shape same as provided image, but single channel
+        '''
+
+        assert 3 == len(rgb_image.shape), \
+            "RGB image array should be 3-dimensional"
+
+        img = self.equalize_histogram(rgb_image)
+        result = self.bilateral_shadow_detection(img)
```

Comment:
może nie `result` a `detected_shadows`? poza tym to teraz bardzo miło się ten algorytm czyta :1st_place_medal: